components/browser/storage-sync/metrics.yaml

```diff
@@ -0,0 +1,257 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+$schema: moz://mozilla.org/schemas/glean/metrics/1-0-0
+
+# The `history_sync` and `bookmarks_sync` metrics use the same structure, but, since are
+# sent in separate pings, after each sync. This means we submit one ping per
+# store per sync.
```

Comment:
It's unfortunate that `history_sync` and `bookmarks_sync` duplicate all the base fields. I thought of having a separate `sync` category, and changing `send_in_pings` to send in both, then send one ping after the other...but the order of individual store (bookmarks and history) syncs within a `StorageSync::sync` operation [isn't guaranteed](https://mozilla.slack.com/archives/CA4JKKLSG/p1559338609037400), and I'm worried that'll interleave counts for both stores.

Reply:
Hmm, so if we have a `metrics.yaml` that looks like this:

```yaml
sync:
  outgoing:
    type: labeled_counter
    labels:
    - uploaded
    - failed_to_upload
  send_in_pings:
  - history_sync
  - bookmarks_sync
```

And we do this:

```kotlin
GleanMetrics.Sync.outgoing["uploaded"].add(historyPingFromRust.uploaded)
GleanMetrics.Pings.historySync.send()
```

And this, potentially before `Pings.historySync.send()` is called:

```kotlin
GleanMetrics.Sync.outgoing["uploaded"].add(bookmarksPingFromRust.uploaded)
GleanMetrics.Pings.bookmarksSync.send()
```

How does Glean know that the second call to `Sync["uploaded"].add()` is meant for bookmarks, not history, if we haven't sent the history ping yet?

Reply:
Hi Lina, jumping in here since @Dexterp37 is already past his EOD.

`history_sync` and `bookmarks_sync` would each have a copy of the same `outgoing` counter.  So in the case you mentioned, both of the stores/pings would increment whenever the `add()` method is called.  In essence, you cannot specifically `add()` to one or the other.  You could end up in situations where the counts were different, depending on when the pings were sent, but you should never be able to add to one and not the other.

In this case, I think what you may want to do is separate them at the metrics level something like this:

```yaml
sync.history:
  outgoing:
    type: labeled_counter
    labels:
    - uploaded
    - failed_to_upload
    send_in_pings:
    - history_sync
sync.bookmarks:
  outgoing:
    type: labeled_counter
    labels:
    - uploaded
    - failed_to_upload
    send_in_pings:
    - bookmarks_sync
```

This keeps the counters separate since they are now treated as unique metrics, and they are then able to be sent in separate pings.  The only difference is now you have two separate metrics that are no longer dependent, and the way you call it changes slightly (notice the additional `.history` in the object hierarchy):
```Kotlin
GleanMetrics.Sync.history.outgoing["uploaded"].add(historyPingFromRust.uploaded)
```

Reply:
Thanks @travis79! Sounds like I'll still need to have the same duplication as in this patch, just nested under `sync`, instead of making them separate top-level metrics.

Reply:
> 
> 
> Thanks @travis79! Sounds like I'll still need to have the same duplication as in this patch, just nested under `sync`, instead of making them separate top-level metrics.

Yes, sorry about the noise, I misunderstood the problem and looks like Travis' solution or your current solution are the correct ones.